File: Segment Tree Modify.cpp
/**
 * Definition of SegmentTreeNode:
 * class SegmentTreeNode {
 * public:
 *     int start, end, max;
 *     SegmentTreeNode *left, *right;
 *     SegmentTreeNode(int start, int end, int max) {
 *         this->start = start;
 *         this->end = end;
 *         this->max = max;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     *@param root, index, value: The root of segment tree and 
     *@ change the node's value with [index, index] to the new given value
     *@return: void
     */
    void modify(SegmentTreeNode *root, int index, int value) {
        // write your code here
        // doModify(root, index, value);
        if (!root || root->start > index || root->end < index)
        {
            return;
        }
        if (root->start == root->end && root->start == index)
        {
            root->max = value;
            return;
        }
        
        modify(root->left, index, value);
        modify(root->right, index, value);
        
        int left_max = root->left == nullptr ? INT_MIN : root->left->max;
        int right_max = root->right == nullptr ? INT_MIN : root->right->max;
        root->max = max(left_max, right_max);
        return;
    }
    int doModify(SegmentTreeNode *root, int index, int value)
    {
        //root nerver should be null.
        if (!root || root->start > index || root->end < index)
        {
            return root->max;
        }
        if (root->start == root->end && root->start == index)
        {
            root->max = value;
            return value;
        }
        else
        {
            int maxValue = doModify(root->left, index, value);
            maxValue = max(maxValue, doModify(root->right, index, value));
            root->max = maxValue;
            return maxValue;
            
        }
    }
};

File: Segment Tree Modify.py
"""
Definition of SegmentTreeNode:
class SegmentTreeNode:
    def __init__(self, start, end, max):
        self.start, self.end, self.max = start, end, max
        self.left, self.right = None, None
"""

class Solution:	
    """
    @param root, index, value: The root of segment tree and 
    @ change the node's value with [index, index] to the new given value
    @return: nothing
    """
    def modify(self, root, index, value):
        # write your code here
        # if not root:
        #     return float('-inf')
        if root.start > index or root.end < index:
            return root.max
        if root.start == root.end == index:
            root.max = value
            return value
        else:
            root.max = (max(self.modify(root.right, index, value), 
                            self.modify(root.left, index, value)))
            return root.max